File: sdks/java/src/test/java/com/hooksniff/PaginationTest.java
package com.hooksniff;

import org.junit.jupiter.api.Test;

import java.util.*;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests for Pagination utility — single page, multi page, empty, max pages, default limit.
 */
public class PaginationTest {

    // ─── Helpers ────────────────────────────────────────────────────────

    /** Build a page response map */
    private static Map<String, Object> page(List<?> data, boolean hasMore) {
        Map<String, Object> m = new HashMap<>();
        m.put("data", data);
        m.put("has_more", hasMore);
        return m;
    }

    // ─── Single page ────────────────────────────────────────────────────

    @Test
    void collectAllSinglePage() {
        List<String> items = List.of("a", "b", "c");
        List<String> result = Pagination.collectAll(
                (limit, offset) -> page(items, false), 10);
        assertEquals(3, result.size());
        assertEquals(List.of("a", "b", "c"), result);
    }

    // ─── Multi page ────────────────────────────────────────────────────

    @Test
    void collectAllMultiplePages() {
        // 3 pages of 2 items each
        List<List<String>> pages = List.of(
                List.of("a", "b"),
                List.of("c", "d"),
                List.of("e")  // last page, has_more = false
        );
        int[] callCount = {0};

        List<String> result = Pagination.collectAll((limit, offset) -> {
            int idx = callCount[0]++;
            if (idx < 2) return page(pages.get(idx), true);
            return page(pages.get(idx), false);
        }, 2);

        assertEquals(5, result.size());
        assertEquals("a", result.get(0));
        assertEquals("e", result.get(4));
        assertEquals(3, callCount[0]);
    }

    // ─── Empty page (no data at all) ───────────────────────────────────

    @Test
    void collectAllEmptyFirstPage() {
        int[] callCount = {0};
        List<String> result = Pagination.collectAll((limit, offset) -> {
            callCount[0]++;
            return page(Collections.emptyList(), false);
        }, 10);

        assertTrue(result.isEmpty());
        assertEquals(1, callCount[0]); // only one call needed
    }

    @Test
    void collectAllNullDataPage() {
        int[] callCount = {0};
        List<String> result = Pagination.collectAll((limit, offset) -> {
            callCount[0]++;
            Map<String, Object> m = new HashMap<>();
            m.put("data", null);
            m.put("has_more", true);
            return m;
        }, 10);

        assertTrue(result.isEmpty());
        assertEquals(1, callCount[0]);
    }

    // ─── Max pages limit ───────────────────────────────────────────────

    @Test
    void collectAllStopsAtMaxPages() {
        int[] callCount = {0};
        List<String> result = Pagination.collectAll((limit, offset) -> {
            callCount[0]++;
            // Always return 1 item with has_more = true — should stop at MAX_PAGES
            return page(List.of("item-" + callCount[0]), true);
        }, 1);

        assertEquals(Pagination.MAX_PAGES, result.size());
        assertEquals(Pagination.MAX_PAGES, callCount[0]);
        assertEquals("item-1", result.get(0));
        assertEquals("item-" + Pagination.MAX_PAGES, result.get(Pagination.MAX_PAGES - 1));
    }

    // ─── Default limit ─────────────────────────────────────────────────

    @Test
    void collectAllDefaultLimitUsesDefaultLimit() {
        int[] capturedLimit = {-1};
        List<String> result = Pagination.collectAll((limit, offset) -> {
            capturedLimit[0] = limit;
            return page(List.of("x"), false);
        });

        assertEquals(Pagination.DEFAULT_LIMIT, capturedLimit[0]);
        assertEquals(1, result.size());
    }

    @Test
    void defaultLimitIs50() {
        assertEquals(50, Pagination.DEFAULT_LIMIT);
    }

    @Test
    void maxPagesIs100() {
        assertEquals(100, Pagination.MAX_PAGES);
    }

    // ─── Offset tracking ───────────────────────────────────────────────

    @Test
    void collectAllTracksOffsetCorrectly() {
        List<Integer> offsets = new ArrayList<>();
        int[] callIdx = {0};
        List<List<String>> pages = List.of(
                List.of("a", "b", "c"),
                List.of("d", "e")
        );

        Pagination.collectAll((limit, offset) -> {
            offsets.add(offset);
            int idx = callIdx[0]++;
            return page(pages.get(idx), idx == 0);
        }, 3);

        assertEquals(0, offsets.get(0));
        assertEquals(3, offsets.get(1)); // offset = previous page size
    }

    // ─── Type generic works with maps ──────────────────────────────────

    @Test
    void collectAllWorksWithMapType() {
        Map<String, Object> item1 = Map.of("id", "ep-1", "url", "https://a.example.com");
        Map<String, Object> item2 = Map.of("id", "ep-2", "url", "https://b.example.com");

        List<Map<String, Object>> result = Pagination.collectAll((limit, offset) ->
                page(List.of(item1, item2), false), 10);

        assertEquals(2, result.size());
        assertEquals("ep-1", result.get(0).get("id"));
        assertEquals("ep-2", result.get(1).get("id"));
    }
}

File: sdks/kotlin/src/test/kotlin/hooksniff/sdk/wrapper/PaginationTest.kt
package hooksniff.sdk.wrapper

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class PaginationTest {

    // ──────────────────────────────────────────────
    // collectAll — single page
    // ──────────────────────────────────────────────
    @Test
    fun `single page returns all items`() {
        val items = listOf("a", "b", "c")
        val result = Pagination.collectAll<String>(limit = 10) { _, _ ->
            Pair(items, false) // hasMore = false
        }
        assertEquals(items, result)
    }

    @Test
    fun `single page with default limit`() {
        // Verify the default limit constant is used when none specified
        var capturedLimit = 0
        Pagination.collectAll<String> { limit, _ ->
            capturedLimit = limit
            Pair(listOf("x"), false)
        }
        assertEquals(Pagination.DEFAULT_LIMIT, capturedLimit)
    }

    // ──────────────────────────────────────────────
    // collectAll — multi page
    // ──────────────────────────────────────────────
    @Test
    fun `multi page collects all items across pages`() {
        val page1 = listOf("a", "b")
        val page2 = listOf("c", "d")
        val page3 = listOf("e")
        var callCount = 0

        val result = Pagination.collectAll<String>(limit = 2) { _, offset ->
            callCount++
            when (offset) {
                0 -> Pair(page1, true)
                2 -> Pair(page2, true)
                4 -> Pair(page3, false)
                else -> Pair(emptyList(), false)
            }
        }
        assertEquals(listOf("a", "b", "c", "d", "e"), result)
        assertEquals(3, callCount)
    }

    @Test
    fun `multi page passes correct offset to fetch`() {
        val offsets = mutableListOf<Int>()

        Pagination.collectAll<String>(limit = 5) { _, offset ->
            offsets.add(offset)
            when (offset) {
                0 -> Pair(listOf("a", "b", "c", "d", "e"), true)
                5 -> Pair(listOf("f"), false)
                else -> Pair(emptyList(), false)
            }
        }
        assertEquals(listOf(0, 5), offsets)
    }

    // ──────────────────────────────────────────────
    // collectAll — empty page
    // ──────────────────────────────────────────────
    @Test
    fun `empty first page returns empty list`() {
        val result = Pagination.collectAll<String>(limit = 10) { _, _ ->
            Pair(emptyList(), true) // hasMore=true but empty data → should break
        }
        assertTrue(result.isEmpty())
    }

    @Test
    fun `empty result when fetch returns empty immediately`() {
        var callCount = 0
        val result = Pagination.collectAll<Int>(limit = 5) { _, _ ->
            callCount++
            Pair(emptyList(), false)
        }
        assertTrue(result.isEmpty())
        assertEquals(1, callCount) // called once, then broke on empty
    }

    // ──────────────────────────────────────────────
    // collectAll — max pages limit
    // ──────────────────────────────────────────────
    @Test
    fun `stops at MAX_PAGES limit`() {
        var callCount = 0
        val result = Pagination.collectAll<Int>(limit = 1) { _, _ ->
            callCount++
            Pair(listOf(callCount), true) // always hasMore
        }
        assertEquals(Pagination.MAX_PAGES, callCount)
        assertEquals(Pagination.MAX_PAGES, result.size)
    }

    @Test
    fun `MAX_PAGES constant is 100`() {
        assertEquals(100, Pagination.MAX_PAGES)
    }

    @Test
    fun `DEFAULT_LIMIT constant is 50`() {
        assertEquals(50, Pagination.DEFAULT_LIMIT)
    }

    // ──────────────────────────────────────────────
    // collectAll — edge cases
    // ──────────────────────────────────────────────
    @Test
    fun `passes limit parameter to fetch function`() {
        var capturedLimit = 0
        Pagination.collectAll<String>(limit = 25) { limit, _ ->
            capturedLimit = limit
            Pair(listOf("a"), false)
        }
        assertEquals(25, capturedLimit)
    }

    @Test
    fun `works with different types`() {
        // Int type
        val ints = Pagination.collectAll<Int>(limit = 3) { _, _ ->
            Pair(listOf(1, 2, 3), false)
        }
        assertEquals(listOf(1, 2, 3), ints)

        // Custom data class
        data class Item(val id: Int, val name: String)
        val items = Pagination.collectAll<Item>(limit = 2) { _, _ ->
            Pair(listOf(Item(1, "a"), Item(2, "b")), false)
        }
        assertEquals(2, items.size)
        assertEquals("a", items[0].name)
    }

    @Test
    fun `last page with hasMore true but empty data stops`() {
        var callCount = 0
        val result = Pagination.collectAll<String>(limit = 2) { _, offset ->
            callCount++
            when (offset) {
                0 -> Pair(listOf("a", "b"), true)
                2 -> Pair(emptyList(), true) // hasMore but empty → break
                else -> Pair(emptyList(), false)
            }
        }
        assertEquals(listOf("a", "b"), result)
        assertEquals(2, callCount)
    }
}
